javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/InstrumentationModuleClassLoader.java

```diff
@@ -0,0 +1,252 @@
+import java.lang.reflect.Method;
+import java.net.URL;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.security.ProtectionDomain;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Classloader used to load the helper classes from {@link
+ * io.opentelemetry.javaagent.extension.instrumentation.InstrumentationModule}s, so that those
+ * classes have access to both the agent/extension classes and the instrumented application classes.
+ *
+ * <p>This classloader implements the following classloading delegation strategy:
+ *
+ * <ul>
+ *   <li>First, injected classes are considered (usually the helper classes from the
+ *       InstrumentationModule)
+ *   <li>Next, the classloader looks in the agent or extension classloader, depending on the where
+ *       the InstrumentationModule comes from
+ *   <li>Finally, the instrumented application classloader is checked for the class
+ * </ul>
+ *
+ * <p>In addition, this classloader ensures that the lookup of corresponding .class resources follow
+ * the same delegation strategy, so that bytecode inspection tools work correctly.
+ */
+public class InstrumentationModuleClassLoader extends ClassLoader {
+
+  private static final Map<String, ClassCopySource> ALWAYS_INJECTED_CLASSES;
+
+  static {
+    Map<String, ClassCopySource> alwaysInjected = new HashMap<>();
+    alwaysInjected.put(
+        LookupExposer.class.getName(), ClassCopySource.create(LookupExposer.class).cached());
+    ALWAYS_INJECTED_CLASSES = Collections.unmodifiableMap(alwaysInjected);
+  }
+
+  private static final ProtectionDomain PROTECTION_DOMAIN = getProtectionDomain();
+  private static final Method FIND_PACKAGE_METHOD = getFindPackageMethod();
+
+  private final Map<String, ClassCopySource> additionalInjectedClasses;
+  private final ClassLoader agentOrExtensionCl;
+  private final ClassLoader instrumentedCl;
+
+  public InstrumentationModuleClassLoader(
+      ClassLoader instrumentedCl,
+      ClassLoader agentOrExtensionCl,
+      Map<String, ClassCopySource> injectedClasses) {
+    // agent/extension-classloader is "main"-parent, but class lookup is overridden
+    super(agentOrExtensionCl);
+    additionalInjectedClasses = injectedClasses;
+    this.agentOrExtensionCl = agentOrExtensionCl;
+    this.instrumentedCl = instrumentedCl;
+  }
+
+  /**
+   * Provides a Lookup within this classloader. See {@link LookupExposer} for the details.
+   *
+   * @return a lookup capable of accessing public types in this classloader
+   */
+  public MethodHandles.Lookup getLookup() {
+    // Load the injected copy of LookupExposer and invoke it
+    try {
+      Class<?> lookupExposer = loadClass(LookupExposer.class.getName());
+      return (MethodHandles.Lookup) lookupExposer.getMethod("getLookup").invoke(null);
```

Comment:
Since this is using actual reflection, is this going to be called often enough that it should be cached/saved somewhere?

Reply:
This will be called once per bootstrapped `invokedynamic`-instruction. So it's not called only once, but called in a rather heavy weight process anyway.
Nonetheless, a cache here won't hurt for future usages.